runtime/pkg/dag/dag_test.go

```diff
@@ -82,3 +85,34 @@ func getTestDAG() *DAG {
 	//     B2  A2
 	return d
 }
+
+func TestCyclicDAG(t *testing.T) {
+	d := NewDAG()
+	d.Add("A0", []string{"B1"})
+	n, err := d.Add("B1", []string{"A0"})
+	require.Error(t, err)
+
+	d.Delete("A0")
+	d.Delete("B1")
```

Comment:
This looks like it will still add `B1` to the DAG even if `d.Add` returns an error? I would expect a function that returns an error to not have side effects in this case.

Reply:
Hey Benjamin, it will return nil
Above test will add B1 due to first Add `d.Add("A0", []string{"B1"})` where we are adding `B1` as parent for `A1`
but it will return `nil` for second `Add` call

Reply:
Okay, got it